Driver for Dell S3048-ON

naved001 opened the pull request:
Closes #844 

* I took the skeleton of the brocade driver and built on top of it. Added some additional helper functions.

* Link to [API guide](http://downloads.dell.com/manuals/all-products/esuprt_ser_stor_net/esuprt_networking/esuprt_net_netwrkng_sw/force10-open-automation_service%20manual9_en-us.pdf) and [CLI Guide](https://topics-cdn.dell.com/pdf/force10-s3048-on_Service%20Manual_en-us.pdf)

* There's one issue that still needs to be solved. I can't remove the Native vlan on this switch, it always defaults to 1 (just like the brocade). To run the deployment tests, I have put in a temporary workaround (it's in the comments) to make sure there are no other issues and it passes both deployment tests.  Any suggestions?

* I am doing some string parsings/manipulation, and if you have better suggestions on how to do that then please let me know.

* I have two of these switches (with all ports available), so if you want to play with it let me know (it would be nice if someone besides me gets to know the switch)

naved001 commented:
oh no, I got @zenhack's name wrong in the last commit :(

naved001 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,392 @@
+        payload = self._make_payload(CONFIG, command)
+        self._make_request('POST', url, data=payload)
+
+    def _remove_all_vlans_from_trunk(self, interface):
```
now that I think of it, I could just call `_remove_vlan_from_trunk` in a loop; because that is exactly what I am doing here.

Or maybe not define this method at all. And do the loop thing in `revert_port` directly.

naved001 commented:
Passing the deployment tests (with the hack)

```
collected 2 items                                                                                                                                                   

tests/deployment/native_networks.py .
tests/deployment/vlan_networks.py .

==================================================================== 2 passed in 326.73 seconds ====================================================================
```

It's quite slow, but not as slow as our pexpect dells

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+            else:
+                self._set_native_vlan(interface, new_network)
+        else:
+            match = re.match(re.compile(r'vlan/(\d+)'), channel)
```
Maybe just call `get_network_allocator().is_legal_channel_for`?

naved001 replied:
This is copy pasta from the brocade code. But yeah, I could change that.

zenhack replied:
In that case, it should probably get changed there as well.

Actually, this whole method feels like boilerplate that should probably get factored out into a super class or something (similarly, the powerconnect and nexus drivers share a lot of structure like this, hence _console.py).

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+            ...
+        }
+
+        """
```
You don't really need to duplicate the docstring from the parent class, and doing so is just going to introduce a hazard of having them get out of sync.

naved001 replied:
Cool. I'll remove it. I just copied this from the brocade.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+        for port in ports:
+            response[port] = filter(None,
+                                    [self._get_native_vlan(port.label)]) \
+                                    + self._get_vlans(port.label)
```
This is a bit convoluted; the use of filter on a one-element list seems like overkill; I think a simple if statement would be clearer.

naved001 replied:
I honestly didn't think much about it because I copied this from the brocade too. I'll make it simpler though.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+        payload = self._make_payload(SHOW, command)
+        response = self._make_request('POST', url, data=payload)
+        # find the Native Vlan ID from the response
+        response = response.text.replace(' ', '')
```
Everything from here up is identical to _get_vlans(); factor this out.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+        begin = response.find('NativeVlanId:') + len('NativeVlanId:')
+        end = response.find('.', begin)
+        vlan = response[begin:end]
+        # a really bad temporary workaround to get the deployment tests to pass
```
Explain what it's doing, why exactly the tests are failing without it, and what's wrong with it.

naved001 replied:
Do you want me to add it as a comment here or explain it to you (because I am pretty sure you already know what's up) ?

zenhack replied:
as a comment, sorry if that wasn't clear.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+        Method relies on the REST API CLI which is slow
+        """
+        if not self._is_port_on(interface):
+            self._port_on(interface)
```
There's a (security sensitive) race-condition here; in the time between turning on the port and setting the right value for the vlan, the node has access to whatever vlan it was on before.

naved001 replied:
I don't think so. A port can only be off if there are no vlans on it.l so when a port is turned on it starts from a default state.

Earlier I thought I could implement revert port by just toggling the port off and on, but the switch doesn't allow that. That's why I added a method to remove all trunked vlans.

zenhack replied:
Is it not on any vlans when it's first turned on? my understanding was we were turning off the port because we couldn't do that. Otherwise, it's on *some* vlan, so unless that's already the one we want to put it on, there is a brief moment where it's on the wrong network. If something goes wrong and we get disconnected from the switch in the middle of this, that leaves us in a situation where that node has illicit access to that network.

naved001 replied:
That's a good point!

You are right, when a switch is first turned on it's starts with a Native Vlan of 1. But I don't really know how to handle that in the driver. A port must be turned on before we do anything to it; and it will start with a default native vlan. 

This seems to be an issue with the [brocade too](https://github.com/CCI-MOC/hil/blob/master/hil/ext/switches/brocade.py#L159)

Doesn't look like we are handling it?

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+        command = 'interface vlan ' + vlan + '\r\n untagged ' + \
+            self.interface_type + ' ' + interface
+        payload = self._make_payload(CONFIG, command)
+        self._make_request('POST', url, data=payload)
```
This should probably get factored out into a helper method; it looks like almost the same copypasta from _get_vlans() and _get_native_vlan().

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+        command = 'interface vlan ' + vlan + '\r\n no untagged ' + \
+            self.interface_type + ' ' + interface
+        payload = self._make_payload(CONFIG, command)
+        self._make_request('POST', url, data=payload)
```
Same thing, helper method.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+        Turn off portmode hybrid, disable switchport, and then shut down the
+        port. All non-default vlans must be removed before calling this.
+
+        Fast method that uses the REST API (no hackery).
```
This doesn't really belong in the docstring, as it's more of an implementation question (same goes elsewhere). If you want to include it, I would do it in a comment just below this, but what I think makes more sense is to put TODO/XXX comments int he places where improvement is desired.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+
+        Returns: string interface
+        """
+        iftypes = {'GigabitEthernet': 'gige-', 'TenGigabitEthernet': 'tengig-',
```
Keep these to one key-value pair per line; having two on some of the lines makes it much harder to read.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+                'interface_type': interface_type,
+                'interface': interface,
+                'suffix': '/%s' % suffix if suffix else ''
+            }
```
Given that you're not actually using anything but %s here, it might make more sense to just use `''.join` instead of a format string.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,387 @@
+        Fast method that uses the REST API (no hackery)
+        """
+
+        url = self._construct_url(interface=port, suffix='\?with-defaults')
```
As far as I can tell this is the only place you're using the suffix argument; why not remove it and just do `+ '\?with-defaults'` here? keeps the complexity out of everything else.

Also: did you really want a backslash there?

naved001 replied:
I wanted to leave it as it is because it might be helpful later when we do some more stuff that might need suffix.

> did you really want a backslash there?

The documentation had that, so I kept it as it is. But you can put any combination of forward and backslashes and it still works.

zenhack replied:
OK. add a comment explaining what's going on there?

naved001 replied:
cool. will do that.

zenhack commented:
>    This seems to be an issue with the brocade too
>
>    Doesn't look like we are handling it?

OK, well we can't just leave this as is (though I might be convinced to
merge this and open a separate issue for the race, since it's the same
bug that we've already got, but see also my comments about factoring out
common logic).  It's not immediately obvious to me what to do about it,
but I will think on the matter.

-Ian

naved001 commented:
> OK, well we can't just leave this as is (though I might be convinced to merge this and open a separate issue for the race, since it's the same bug that we've already got, but see also my comments about factoring out common logic). It's not immediately obvious to me what to do about it, but I will think on the matter.

I had a chat with @okrieg and he proposed a solution (similar to  #852 ):

HIL should enforce that a node can only be powered on when a native vlan has been set. 
So when a user gets a node, they must first set a native vlan. The call to `node_power_on` should ensure that a native vlan has been set on all connected nics (make a call to the switch to get the native vlan) before it actually powers on the node.

The idea seems simple enough, if we agree on this we could think about the implementation details. 

@henn curious to your thoughts on this as well.

zenhack commented:
That sounds like a basically sane solution to me. I'm a little worried about the proliferation of stateful safety constraints in our code, since it's going to make implementation more hazardous.

naved001 commented:
I think I have factored out what I thought was reasonable and addressed the comments you made.

Re: using `is_legal_channel_for()`, I made that changed but now this happened when I was running deployment tests.
```
self = <hil.ext.network_allocators.vlan_pool.VlanAllocator object at 0x7f141919af10>, net_id = None

    def legal_channels_for(self, net_id):
        return ["vlan/native",
>               "vlan/" + net_id]
E       TypeError: cannot concatenate 'str' and 'NoneType' objects
```

Before I fix that, I am curious why are we doing this check again in the driver? We already do it in the [API here](https://github.com/CCI-MOC/hil/blob/master/hil/api.py#L419)

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,377 @@
+        response = {}
+        for port in ports:
+            native = self._get_native_vlan(port.label)
+            # return an empty list if native is None
```
Better to just change the expression to make this more obvious; options:

* `native = [native] if native is not None else []`
* `native = [] if native is None else [native]`
* An if-statement, rather than expression:

```python
if native is None:
    native = []
else:
    native = [native]
```

zenhack replied:
Better yet, you could just do:

```python
response[port] = self._get_vlans(port.label)
if native is not None:
    response[port].append(native)
```

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,377 @@
+        # The switchport defaults to native vlan 1. Our deployment tests make
+        # assertions that a switchport has no default vlan at start.
+        # To get the tests to pass, we return None if we see the switchport has
+        # a default native vlan (=1). This needs to be fixed ASAP.
```
maybe put the work FIXME in here, just so it's easier to grep for.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,377 @@
+        i-Internaluntagged,I-Internaltagged,v-VLTuntagged,V-VLTtagged\r\n\r\n
+        Name:GigabitEthernet1/3\r\n802.1QTagged:Hybrid\r\nVlanmembership:\r\n
+        QVlans\r\nU1512\r\n T1511 \r\n\r\n NativeVlanId:1512 .\r\n\r\n\r\n\r\n
+        MOC-Dell-S3048-ON#</command>\n</output>\n"
```
You could probably format this in a way that's easier to read, e.g:

```
u"<output xmlns='http://www.dell.com/ns/dell:0.1/root'>\n" \
    "<command>showinterfacesswitchportGigabitEthernet1/3\r\n" \
    "\r\n" \
    "Codes:U-Untagged,..." \
    ...
```

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,377 @@
+        1. we are using the REST API CLI
+        2. we have to go to every vlan and then remove our port from it;
+           that makes multiple calls.
+        """
```
I don't think we should do it right now, but question: is it possible to parallelize this, or will the switch choke on that? I suspect we can improve on the latency if we have more than one request in flight at a time. Also, how slow are we talking?

naved001 replied:
I think it's doable within this PR. 

This is what `_remove_native_vlan` does:

```python
        command = 'interface vlan ' + vlan + '\r\n no tagged ' + \
            self.interface_type + ' ' + interface
```
that translates to 2 commands in one call to the switch. The commands are separated by \r\n
```
switch(conf)# interface vlan vlanno
switch(conf-if-vl-vlanno)# no tagged gigabitethernet 1/3
```

So for removing multiple vlans, I should be able to generate a huge command (separated by \r\n) and send that in one call. That would probably be faster.

> Also, how slow are we talking?

Both deployment tests take around 340 seconds to complete.

naved001 replied:
So, I did this. The deployment tests don't seem to be any faster. But I guess the speedup is proportional to part of program that can be parallelized.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,377 @@
+        response = self._make_request('GET', url)
+        root = etree.fromstring(response.text)
+        shutdown = root.find(self._construct_tag('shutdown')).text
+        # is error handling needed here?
```
Probably for robustness, we should do something like:

```python
if shutdown not in ('true', 'false'):
    raise SomeKindOfError()
return shutdown == 'false'
```

...rather than treating possibly bogus data as the port being off, which could result in various information leaks.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,377 @@
+    @staticmethod
+    def _convert_interface(interface_type):
+        """ Convert the interface name from switch CLI form to what the API
+        server understands.
```
This docstring doesn't look like it's quite right -- you're converting interface *types*, not full names of specific interfaces, unless I'm mistaken.

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,377 @@
+        return '{http://www.dell.com/ns/dell:0.1/root}%s' % name
+
+    def _make_request(self, method, url, data=None,
+                      acceptable_error_codes=()):
```
afaik, you don't use this parameter anywhere. More copypasta from the brocade driver.

In the future, please don't copy-and-paste; if there's enough in common to warrant it, there's enough in common to factor it out into its own code.

zenhack commented:
This is much better. I left a number of other comments, now that it's easier to follow.

zenhack commented:
Quoting Naved Ansari (2017-09-12 10:22:59)

>    So for removing multiple vlans, I should be able to generate a huge
>    command (separated by \r\n) and send that in one call. That would
>    probably be faster.

Yeah, that would avoid doing so many round-trips. Sure, give it a shot.

naved001 commented:
@zenhack I think addressed all the comments. 

@shwsun @SahilTikale can we get a second review here? Please ask any questions you have. Would be nice to have another exhaustive review.

geedo0 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,380 @@
+# Copyright 2017 Massachusetts Open Cloud Contributors
```
Do you need to consider renaming the hil.ext.switches.dell module to something more explicit now?

geedo0 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,380 @@
+        response = self._get_port_info(interface).splitlines()
+
+        try:
+            index = response.index("Vlanmembership:") + 3
```
The 3 here is kind of troubling. Can we add some more checks here to make sure you're parsing out the right thing? You can also do a tighter regex to match exactly the line with the "T". Bounds checking on the array would be nice as well to avoid an out of bounds exception.

naved001 replied:
Good call,  I should have been using regex all along.

geedo0 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,380 @@
+        response = self._get_port_info(interface)
+        if response.find('NativeVlanId:') == -1:
+            return None
+        begin = response.find('NativeVlanId:') + len('NativeVlanId:')
```
Use a regular expression to figure out if the native vlan exists and to get its id.

r'NativeVlanId:(\d+)\.'

geedo0 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,380 @@
+        """
+        if not self._is_port_on(interface):
+            self._port_on(interface)
+        command = 'interface vlan ' + vlan + '\r\n tagged ' + \
```
This driver constructs enough commands with the \r\n line endings that I think it's worthwhile to use a different pattern to reduce the bugs.

Something like this might work well
```
self._build_cmd = lambda cmd: '\r\n'.join(cmd)
```

naved001 replied:
Could you explain how this reduces bugs? I avoid using lambas because they make the code slightly less readable imo.

jeremyfreudberg replied:
sorry to butt in, but please do not assign names to anonymous functions. if you want a named function use def.

naved001 replied:
@jeremyfreudberg feel free to butt in.

geedo0 replied:
Yeah, I was just being really lazy. My main point though was to define a function that would build these command sequences with the /r/n instead of constructing them inline. It would take in an array of commands and then return the string delimited by windows newlines.

geedo0 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,380 @@
+        Args:
+            interface: interface to remove the native vlan from.vlan
+        """
+        vlan = self._get_native_vlan(interface)[1]
```
This will throw a TypeError if _get_native_vlan returns None

geedo0 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,380 @@
+        if interface is None:
+            return '%s/api/running/dell/_operations/cli' % self.hostname
+
+        val = re.compile(r'^\d+/\d+(/\d+)?$')
```
You can call validate_port_name(interface) here instead to improve reuse.

geedo0 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,380 @@
+        """Valid port names for this switch are of the form 1/0/1 or 1/2"""
+
+        val = re.compile(r'^\d+/\d+(/\d+)?$')
+        if not val.match(port):
```
This syntax is cleaner since the regex pattern isn't reused.
```
if not re.match(r'^\d+/\d+(/\d+)?$', port):
```

zenhack commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,380 @@
+# Copyright 2017 Massachusetts Open Cloud Contributors
```
Yeah, at some point we should re-name it to powerconnect. That gets into some not-entirely trivial issues with migrating existing setups though, so my inclination is to defer and open a separate issue.

zenhack commented on docs/network-drivers.md:
```diff
@@ -213,6 +213,40 @@ The body of the api call request will look like:
+"GigabitEthernet", "TenGigabitEthernet". Please look at the switch guide for
+more valid interface types.
+
+The API server either runs on port 8008 (HTTP) or 8888 (HTTPS), so be sure to
```
I would say the switch's API server to make this a bit more clear -- it took me a second to figure out you weren't talking about the HIL API server, and I was momentarily confused.

zenhack commented on docs/network-drivers.md:
```diff
@@ -213,6 +213,40 @@ The body of the api call request will look like:
+specify that in the ``hostname``.
+
+If you have multiple types of ports on the same switch, register the switch
+multiple times with different parameters for ``interface_type``.
```
This is kindof gross, is it really much work to do this as port_register?

naved001 replied:
I thought this was reasonable and generally accepted since I got this from the brocade driver.

Doing this in port register means, I would have to read the whole interface_type + interface like `TenGigabitEthernet 1/3`. 

Anyway, making this change would require a significant amount of changes everywhere.

zenhack replied:
Ug, okay, I guess we can live with it. I'm realizing I'm somewhat dissatisfied with the quality of the brocade driver...

naved001 replied:
yeah, I'll avoid copy-pasting in the future.

naved001 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,380 @@
+        return
+
+    def disconnect(self):
+        pass
```
I added a docstring to explain this.

In the base [Switch class](https://github.com/CCI-MOC/hil/blob/master/hil/model.py#L243) we explain that for switches that are not connection-oriented, we don't need to worry about setting up and/or disconnecting a session.

naved001 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -0,0 +1,380 @@
+            vlan_id = channel.replace('vlan/', '')
+            legal = get_network_allocator(). \
+                is_legal_channel_for(channel, vlan_id)
+            assert legal, "HIL passed an invalid channel to the switch!"
```
True, it actually needs to be changed in the brocade too. That will be done in a separate PR though.

naved001 commented:
@geedo0 @zenhack 

pushed another commit addressing your changes.

geedo0 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -153,11 +147,9 @@ def _get_native_vlan(self, interface):
-        end = response.find('.', begin)
-        vlan = response[begin:end]
+        match = re.search(re.compile(r'NativeVlanId:(\d+)\.'), response)
+        vlan = re.search(re.compile(r'(\d+)'), match.group())
```
We lost handling for the error condition where the NativeVlanId was not found. Also, the vlanid is already captured in the original regex so something like this should work.

```
match = re.search(r'NativeVlanId:(\d+)\.', response)
if match is not None:
    vlan = match.group(1)
else:
    return None
```

naved001 replied:
oh, I totally forgot about that haha. Good catch.

And, I was doing `match.group()[1]` which was just returning 'a'.

geedo0 commented on hil/ext/switches/dellnos9.py:
```diff
@@ -128,18 +128,12 @@ def _get_vlans(self, interface):
-            # a port with no tagged interface has no section with "T"
+        response = self._get_port_info(interface)
+        # finds a comma separated list of integers starting with "T"
+        match = re.search(re.compile(r'T(\d+)((,\d+)?)*'), response)
```
You don't need the re.compile(), search will accept the regex string directly :)

naved001 replied:
@geedo0 Done!

zenhack commented:
LGTM, merging.
